### ISSUE TITLE
(Streetcar) vehicles do not use all open stations/gates

### ISSUE THREAD

**Issue by LubosKolouch** [OpenTTD/OpenTTD#8101]
## Version of OpenTTD
1.10.1

## Expected result
(Streetcar) Vehicles using all empty stations/gates

## Actual result
(Streetcar) vehicles queue before full station/gate, even though there are empty ones.

## Steps to reproduce
1. Create bigger (streetcar) station
2. Generate traffic to/from the station
3. See that the vehicles don't utilize all open station parts

**Comment by LubosKolouch** [comment]
![ottd_issue](https://user-images.githubusercontent.com/4145223/80520345-944fbd00-8989-11ea-8ffc-8e3eebe9f094.png)

**Comment by Yexo** [comment]
Can you upload a (small) savegame that reproduces this?

**Comment by LubosKolouch** [comment]
[save.zip](https://github.com/OpenTTD/OpenTTD/files/4547697/save.zip)

See vehicle #10

**Comment by LubosKolouch** [comment]
![Screenshot at 2020-04-28 20-09-35](https://user-images.githubusercontent.com/4145223/80522042-383a6800-898c-11ea-9f39-a43f9cc7637c.png)

**Comment by LubosKolouch** [comment]
Is there something else I can do to clarify this issue more?

**Comment by LordAro** [comment]
This is due to the road vehicle pathfinder cache. It caches the whole route, rather than most of the route (start and end a few tiles away from the stations). This results in this behaviour with multi-tile stops

I could've sworn there was already an issue about this, but can't find one. Very similar to #7670

**Comment by JGRennison** [comment]
The cache issue should be fixed as of #7979.

> [save.zip](https://github.com/OpenTTD/OpenTTD/files/4547697/save.zip)
> 
> See vehicle #10

Vehicle number 10 does not have proper orders. Giving it orders seems to make the problem go away.

**Comment by LubosKolouch** [comment]
> The cache issue should be fixed as of #7979.
> 
> 
> 
> > [save.zip](https://github.com/OpenTTD/OpenTTD/files/4547697/save.zip)
> 
> > 
> 
> > See vehicle #10
> 
> 
> 
> Vehicle number 10 does not have proper orders. Giving it orders seems to make the problem go away.

I'm sorry but it's not the main issue, it's probably the the comment above. It happens even when vehicle has proper orders.

**Comment by LordAro** [comment]
I cannot reproduce this issue any longer. If you can, please upload a demonstrator save game (ideally, after trying it with 1.11.0-beta1 or a recent nightly)

![image](https://user-images.githubusercontent.com/4007992/108643780-ef681880-74a3-11eb-885b-04fbe48afb93.png)

**Comment by LubosKolouch** [comment]
The issue is still there with 1.11.0 - Vehicle 10 keeps selecting occupied docks, even though there are empty ones (even if you reverse it...)
![Screenshot_2021-04-12_12-56-21](https://user-images.githubusercontent.com/4145223/114384050-855c1d80-9b8e-11eb-9251-681ffad576c1.png)

[Marningville Transport, Oct 29th, 2000.sav.zip](https://github.com/OpenTTD/OpenTTD/files/6296002/Marningville.Transport.Oct.29th.2000.sav.zip)

**Comment by glx22** [comment]
On screenshot I can see vehicle 10 has no orders.

**Comment by LubosKolouch** [comment]
![image](https://user-images.githubusercontent.com/4145223/114398218-d4f71500-9b9f-11eb-8ec6-08042fa76833.png)
which is strange, what means implicit order?

**Comment by nielsmh** [comment]
Implicit orders are for stations the vehicle passed through moving along the track/road, and stopped at because it didn't have a non-stop order. Implicit orders are collected to support the cargo distribution network graph (even when cargo distribution is disabled), but they do not order the vehicle to go anywhere. Rather, they are an indication of where the vehicle ended up going.

**Comment by LubosKolouch** [comment]
OK. I've given the vehicle order, reversed it... and guess where is it going?
![image](https://user-images.githubusercontent.com/4145223/114453118-82d2e580-9bd9-11eb-858c-bdde059e38f0.png)

**Comment by LubosKolouch** [comment]
and before someone mentions it - yes, I realize this is a corner case. But it demonstrates the problem well - (at least) streetcars do not choose correctly empty stations, but instead wait in front of a full one... (blocking all traffic)

**Comment by LC-Zorg** [comment]
I know this problem, or actually two, and I must say that this presentation is terrible. It neither describes nor shows anything, apart from the author's basic mistakes, such as the lack of orders or the expectation that the tram will turn back at an intersection.

Two problems that exist here (not only for trams):
1. Pathfinder sending a vehicle to a stop too late marks this stop as occupied. Then when several vehicles are going one after the other lead to them all going to the same stop even if there are many others available.

2. When all stops are full, the vehicle is steered to the closest bus stop which is occupied. Usually, it is the other stop is released first, and the vehicle that didn't take place is now standing on the road and blocking it.

I have rebuilt that save:
- I withdrawn all the trams to the depot, gave them the right orders, released them and... I don't see any problem.
- I created loops to have vehicles repeatedly stopping and... still, as long as there are not more vehicles than stops, the problem does not occur.
- The situation changes when buying another tram - this causes the emergence of the two problems mentioned above. Both are additionally shown separately in the save.

![Problems with road stops](https://user-images.githubusercontent.com/67662046/114633456-cb190300-9cc0-11eb-8ec1-64d957ed116e.png)
[Marningville Transport - where is the problem.zip](https://github.com/OpenTTD/OpenTTD/files/6307217/Marningville.Transport.-.where.is.the.problem.zip)
